File: 1657-find-the-winner-of-an-array-game/1657-find-the-winner-of-an-array-game.java
class Solution {
    public int getWinner(int[] arr, int k) {
        Queue<Integer> q = new LinkedList<>();
        int max = 0;
        for(int num : arr){
            max = Math.max(num, max);
            q.add(num);
        }
        int curr = q.poll();
        int streak = 0;
        while(true){
            int opp = q.poll();

            if(curr > opp){
                q.add(opp);
                streak++;
            }
            else{
                q.add(curr);
                curr = opp;
                streak = 1;
            }

            if(streak == k || curr == max){
                return curr;
            }
        }
    }
}

File: 1657-find-the-winner-of-an-array-game/1657-find-the-winner-of-an-array-game.py
class Solution:
    def getWinner(self, arr: List[int], k: int) -> int:
        q = deque(arr)
        max_val = max(arr)
        curr = q.popleft()
        streak = 0
        while True:
            opp = q.popleft()
            if curr > opp:
                q.append(opp)
                streak += 1
            else:
                q.append(curr)
                curr = opp
                streak = 1

            if streak == k or curr == max_val:
                return curr